envoy-control-tests/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/ssl/EnvoySANValidationTest.kt

```diff
@@ -0,0 +1,74 @@
+package pl.allegro.tech.servicemesh.envoycontrol.ssl
+
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.BeforeAll
+import org.junit.jupiter.api.Test
+import pl.allegro.tech.servicemesh.envoycontrol.config.Ads
+import pl.allegro.tech.servicemesh.envoycontrol.config.EnvoyControlRunnerTestApp
+import pl.allegro.tech.servicemesh.envoycontrol.config.EnvoyControlTestConfiguration
+import pl.allegro.tech.servicemesh.envoycontrol.config.Xds
+
+class AdsEnvoySANValidationTest : EnvoySANValidationTest() {
+
+    companion object {
+        private val properties = mapOf(
+                "envoy-control.envoy.snapshot.trustedCaFile" to "/usr/local/share/ca-certificates/root-ca.crt"
+        )
+
+        @JvmStatic
+        @BeforeAll
+        fun setupTest() {
+            setup(
+                envoyConfig = Ads,
+                appFactoryForEc1 = { consulPort -> EnvoyControlRunnerTestApp(properties, consulPort) }
+            )
+        }
+    }
+}
+
+class XdsEnvoySANValidationTest : EnvoySANValidationTest() {
+    companion object {
+        private val properties = mapOf(
+                "envoy-control.envoy.snapshot.trustedCaFile" to "/usr/local/share/ca-certificates/root-ca.crt"
+        )
+
+        @JvmStatic
+        @BeforeAll
+        fun setupTest() {
+            setup(
+                envoyConfig = Xds,
+                appFactoryForEc1 = { consulPort -> EnvoyControlRunnerTestApp(properties, consulPort) }
+            )
+        }
+    }
+}
+
+abstract class EnvoySANValidationTest : EnvoyControlTestConfiguration() {
+
+    @Test
+    fun `should validate SAN part of certificate`() {
+        // given
+        envoyContainer1.addHost("my.example.com", httpsEchoContainer.ipAddress())
+        envoyContainer1.addHost("bad.host.example.com", httpsEchoContainer.ipAddress())
+
+        untilAsserted {
+            // when
+            val reachableResponse = callDomain("my.example.com")
+
+            assertThat(reachableResponse).isOk()
+        }
+    }
+
+    @Test
+    fun `should reject certificate without matching SAN`() {
+        // given
+        envoyContainer1.addHost("bad.host.example.com", httpsEchoContainer.ipAddress())
+
+        untilAsserted {
+            // when
+            val reachableResponse = callDomain("bad.host.example.com")
+
+            assertThat(reachableResponse).isUnreachable()
```

Comment:
I am worried that this test doesn't actually verify that invalid certificate is rejected. It will pass at first iteration, when the target cluster is simply not available yet.

We should find a way to verify that envoy rejected a certificate. Maybe envoy returns different status code or message? If not, maybe we should check a metric that counts invalid certificate events?

Or simply we should merge two tests - first check if a good certificate is accepted and *then* check that an invalid certificate is rejected.